digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/csvDataImporter/TrafficLightsCsvImporter.scala

```diff
@@ -15,24 +18,201 @@ class TrafficLightsCsvImporter(roadLinkServiceImpl: RoadLinkService, eventBusImp
-    pointAssetAttributes.foreach { trafficLightAttribute =>
+    val trafficLights = pointAssetAttributes.map { trafficLightAttribute =>
       val csvProperties = trafficLightAttribute.properties
       val nearbyLinks = trafficLightAttribute.roadLink
-
+      val optBearing = tryToInt(getPropertyValue(csvProperties, "trafficLightBearing").toString)
       val position = getCoordinatesFromProperties(csvProperties)
 
-      val roadLink = roadLinkService.enrichRoadLinksFromVVH(nearbyLinks)
-      val nearestRoadLink = roadLink.filter(_.administrativeClass != State).minBy(r => GeometryUtils.minimumDistance(position, r.geometry))
+      val (assetBearing, assetValidityDirection) = trafficLightsService.recalculateBearing(optBearing)
+
+      val possibleRoadLinks = roadLinkService.filterRoadLinkByBearing(assetBearing, assetValidityDirection, position, nearbyLinks)
+
+      val roadLinks = possibleRoadLinks.filter(_.administrativeClass != State)
+      val roadLink = if (roadLinks.nonEmpty) {
+        roadLinks.minBy(r => GeometryUtils.minimumDistance(position, r.geometry))
+      } else
+        nearbyLinks.minBy(r => GeometryUtils.minimumDistance(position, r.geometry))
+
+      val validityDirection = if(assetBearing.isEmpty) {
+        trafficLightsService.getValidityDirection(position, roadLink, assetBearing)
+      } else assetValidityDirection.get
+
+      val mValue = GeometryUtils.calculateLinearReferenceFromPoint(position, roadLink.geometry)
+
+      val sidecodeProperty = AssetProperty("trafficLightSidecode", validityDirection)
+
+      (csvProperties, CsvPointAsset(position.x, position.y, roadLink.linkId, generateBaseProperties(csvProperties ++ Seq(sidecodeProperty)), validityDirection, assetBearing, mValue, roadLink, (roadLinks.isEmpty || roadLinks.size > 1) && assetBearing.isEmpty))
+    }
 
-      val floating = checkMinimumDistanceFromRoadLink(position, nearestRoadLink.geometry)
+    val notImportedData: Seq[NotImportedData] = trafficLights.map { case (csvRow, trafficLight) =>
+      val mValue = GeometryUtils.calculateLinearReferenceFromPoint(Point(trafficLight.lon, trafficLight.lat), trafficLight.roadLink.geometry)
+      val bearing = if(trafficLight.bearing.isEmpty && !trafficLight.isFloating)
+        Some(GeometryUtils.calculateBearing(trafficLight.roadLink.geometry, Some(mValue)))
+      else
+        trafficLight.bearing
+      try {
+        val trafficLightsSuitableForMerge = trafficLightsService.fetchSuitableForMergeByRadius(centralPoint = Point(trafficLight.lon, trafficLight.lat), radius = 5)
+        if (trafficLightsSuitableForMerge.isEmpty)
+          trafficLightsService.createFromCoordinates(IncomingTrafficLight(trafficLight.lon, trafficLight.lat, trafficLight.roadLink.linkId, trafficLight.propertyData, Some(trafficLight.validityDirection), bearing), trafficLight.roadLink, user.username, trafficLight.isFloating)
```

Comment:
The central point for this verification is starting from the asset to be created, if there is another traffic light close and in suitable condition for the merge it will merge the new one into the existing one. Otherwise it will create a new one on the coordinates of the new traffic light to be created

Reply:
Ok, I understand.